drivers/usb/usbstor_new/fdo.c

```diff
@@ -459,3 +486,155 @@ USBSTOR_FdoHandlePnp(
+    PFDO_DEVICE_EXTENSION FdoExtension;
+    PIRP CurrentPowerIrp;
+
+    FdoDevice = Context;
+    FdoExtension = FdoDevice->DeviceExtension;
+
+    CurrentPowerIrp = FdoExtension->CurrentPowerIrp;
+    FdoExtension->CurrentPowerIrp = NULL;
+
+    PoStartNextPowerIrp(CurrentPowerIrp);
+
+    IoCopyCurrentIrpStackLocationToNext(CurrentPowerIrp);
+    IoMarkIrpPending(CurrentPowerIrp);
+
+    PoCallDriver(FdoExtension->LowerDeviceObject, CurrentPowerIrp);
+}
+
+NTSTATUS
+NTAPI
+USBSTOR_FdoSetD0Completion(
+    IN PDEVICE_OBJECT DeviceObject,
+    IN PIRP Irp,
+    IN PVOID Context)
+{
+    NTSTATUS Status;
+    KIRQL OldIrql;
+
+    Status = Irp->IoStatus.Status;
+
+    KeRaiseIrql(DISPATCH_LEVEL, &OldIrql);
+    IoStartNextPacket(DeviceObject, TRUE);
+    KeLowerIrql(OldIrql);
+
+    PoStartNextPowerIrp(Irp);
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+USBSTOR_FdoSetPower(
+    IN PDEVICE_OBJECT DeviceObject,
+    IN PIRP Irp)
+{
+    PFDO_DEVICE_EXTENSION FdoExtension;
+    PIO_STACK_LOCATION IoStack;
+    POWER_STATE_TYPE Type;
+    POWER_STATE State;
+    POWER_STATE NewState;
+    POWER_STATE OldState;
+
+    DPRINT("USBSTOR_FdoSetPower: DeviceObject %p, Irp %p\n", DeviceObject, Irp);
+
+    FdoExtension = DeviceObject->DeviceExtension;
+
+    IoStack = IoGetCurrentIrpStackLocation(Irp);
+    Type = IoStack->Parameters.Power.Type;
+    State = IoStack->Parameters.Power.State;
+
+    switch (Type)
+    {
+        case SystemPowerState:
+        {
+            FdoExtension->SystemState = State.SystemState;
+
+            if (State.SystemState == PowerSystemWorking)
+                NewState.DeviceState = PowerDeviceD0;
+            else
+                NewState.DeviceState = PowerDeviceD3;
+
+            if (FdoExtension->DeviceState != NewState.DeviceState)
+            {
+                FdoExtension->CurrentPowerIrp = Irp;
+
+                DPRINT("USBSTOR_FdoSetPower: State.SystemState - %x\n",
+                       State.SystemState);
+
+                return PoRequestPowerIrp(FdoExtension->PhysicalDeviceObject,
+                                         IRP_MN_SET_POWER,
+                                         NewState,
+                                         USBSTOR_FdoSetPowerCompletion,
+                                         DeviceObject,
+                                         NULL);
+            }
+
+            break;
+        }
+        case DevicePowerState:
+        {
+            OldState.DeviceState = FdoExtension->DeviceState;
+            FdoExtension->DeviceState = State.DeviceState;
+
+            DPRINT("USBSTOR_FdoSetPower: State.DeviceState - %x, OldState.DeviceState - %x\n",
+                   State.DeviceState,
+                   OldState.DeviceState);
+
+            if (OldState.DeviceState == PowerDeviceD0)
+            {
+                if (State.DeviceState > PowerDeviceD0)
+                {
+                    ULONG Key = 0;
+                    IoStartPacket(DeviceObject, Irp, &Key, NULL);
```

Comment:
A naïve question: where is the Irp inserted in the queue if &Key is replaced to NULL? (this doesn't seem to be mentioned in MSDN).

Reply:
Also, depending on whether this is "common" knowledge or not, perhaps adding a comment why using this "Key" value ("insert the Irp at the tail of the queue") would be welcome.

Reply:
Maybe I did not understand the question correctly, but:
https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/ntifs/nf-ntifs-iostartpacket :
Key 
Pointer to a value that determines where to insert the packet into the device queue. If this is zero, the packet is inserted at the tail of the device queue.